server/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -1296,14 +1296,20 @@ private <T, C extends Collection<? super T>> C readCollection(Writeable.Reader<T
 
     /**
      * Reads an enum with type E that was serialized based on the value of its ordinal
+     * @deprecated give each {@link Writeable} enum an {@code byte id} member
+     *             and serialize that
      */
+    @Deprecated
     public <E extends Enum<E>> E readEnum(Class<E> enumClass) throws IOException {
```

Comment:
We would need to deprecate `readEnumSet` and `writeEnumSet` as well.

Reply:
:+1: although EnumSet might be useful somethings, it might be worth providing `readEnumSet` alternative that accepts a reader.